### PR TITLE
Makes some machines upgradeable

### PR DESCRIPTION
**Pull request by printer16**
Adds the ability to upgrade the following

- Sleepers
- The ore processor
- All cooking appliances (Minus the microwave)
- Tesla Coil

Also makes components not spawn in machines until they are needed (When they are being upgraded) to reduce the objects generated. Adds some new macro's and converts some istypes to use them. And moves stock parts into their own file.

### CODE REVIEW

File: code/defines/obj/weapon.dm [hunk 1]
```diff
@@ -315,19 +315,21 @@
...
 	desc = "An electronic module for reading data and ID cards."
 
-/obj/item/weapon/module/power_control
+/obj/item/weapon/circuitboard/power_control
 	name = "power control module"
 	icon_state = "power_mod"
 	desc = "Heavy-duty switching circuits for power control."
 	matter = list(DEFAULT_WALL_MATERIAL = 50, "glass" = 50)
+	req_components = list(
+						"/obj/item/weapon/stock_parts/capacitor" = 4,
```
Comment (Lohikar) [inline code/defines/obj/weapon.dm]: This should be a type, not a string.

File: code/game/machinery/machinery.dm [hunk 2]
```diff
@@ -129,24 +129,11 @@ Class Procs:
 	var/tmp/machinery_processing = FALSE	// Are we process()ing in SSmachinery?
 	var/has_special_power_checks = FALSE	// If true, call auto_use_power instead of doing it all in SSmachinery.
 
-/obj/machinery/Initialize(mapload, d = 0, populate_components = TRUE)
+/obj/machinery/Initialize(mapload, d = 0, populate_components = FALSE)
```
Comment (Lohikar) [inline code/game/machinery/machinery.dm]: Remove this argument if it isn't being used.

File: code/game/machinery/vending.dm [hunk 3]
```diff
@@ -97,6 +99,14 @@
 	
 	var/can_move = 1	//if you can wrench the machine out of place
 
+	component_types = list(
```
Comment (Lohikar) [inline code/game/machinery/vending.dm]: Why are vendors upgradable? The delays are for the animations on some types, it shouldn't change.

File: code/game/objects/items/weapons/circuitboards/machinery/miscboards.dm [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+	origin_tech = list(TECH_DATA = 4, TECH_BIO = 2)
+	board_type = "other"
+
+/obj/item/weapon/circuitboard/vending
+	name = T_BOARD("vending machine")
+	desc = "The circuitboard for a vending machine. Not of much use."
+	origin_tech = list(TECH_MAGNET = 2, TECH_ENGINEERING = 2)
+	board_type = "other"
+	req_components = list(
+							"/obj/item/weapon/stock_parts/capacitor" = 2,
```
Comment (Lohikar) [inline code/game/objects/items/weapons/circuitboards/machinery/miscboards.dm]: These should be types.

File: code/game/objects/items/weapons/circuitboards/machinery/miscboards.dm [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+							"/obj/item/weapon/stock_parts/matter_bin" = 1,
+							"/obj/item/weapon/stock_parts/console_screen" = 1)
+
+/obj/item/weapon/circuitboard/sleeper
+	name = T_BOARD("sleeper")
+	desc = "The circuitboard for a sleeper."
+	origin_tech = list(TECH_MAGNET = 2, TECH_ENGINEERING = 2)
+	board_type = "other" // change this to machine if you want it to be buildable
+	req_components = list(
+							"/obj/item/weapon/stock_parts/capacitor" = 2,
```
Comment (Lohikar) [inline code/game/objects/items/weapons/circuitboards/machinery/miscboards.dm]: And these.

File: code/game/objects/items/weapons/circuitboards/machinery/miscboards.dm [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+							"/obj/item/weapon/stock_parts/scanning_module" = 2,
+							"/obj/item/weapon/stock_parts/console_screen" = 1,
+							"/obj/item/weapon/reagent_containers/glass/beaker/large" = 1)
+
+/obj/item/weapon/circuitboard/refiner
+	name = T_BOARD("ore processer")
```
Comment (Lohikar) [inline code/game/objects/items/weapons/circuitboards/machinery/miscboards.dm]: Processor.

File: code/game/objects/items/weapons/circuitboards/machinery/miscboards.dm [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+	desc = "The circuitboard for an ore processing machine."
+	origin_tech = list(TECH_MAGNET = 2, TECH_ENGINEERING = 2)
+	board_type = "other" // change this to machine if you want it to be buildable
+	req_components = list(
+							"/obj/item/weapon/stock_parts/capacitor" = 2,
```
Comment (Lohikar) [inline code/game/objects/items/weapons/circuitboards/machinery/miscboards.dm]: More types as strings.

File: code/game/objects/items/weapons/circuitboards/machinery/miscboards.dm [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+							"/obj/item/weapon/stock_parts/matter_bin" = 1,
+							"/obj/item/weapon/stock_parts/micro_laser" = 2)
+
+/obj/item/weapon/circuitboard/cooking
+	name = T_BOARD("kitchen appliance")
+	desc = "The circuitboard for many kitchen appliances. Not of much use."
+	origin_tech = list(TECH_MAGNET = 2, TECH_ENGINEERING = 2)
+	board_type = "other"
+	req_components = list(
+							"/obj/item/weapon/stock_parts/capacitor" = 3,
```
Comment (Lohikar) [inline code/game/objects/items/weapons/circuitboards/machinery/miscboards.dm]: Same.

File: code/modules/power/apc.dm [hunk 9]
```diff
@@ -120,6 +121,12 @@
 
 	var/datum/effect_system/sparks/spark_system
 
+	component_types = list(
```
Comment (Lohikar) [inline code/modules/power/apc.dm]: Aren't these dropped on deconstruction, but not required for construction? Won't this allow for infinite power_control boards, capacitors, and console_screens?

File: code/modules/power/tesla/coil.dm [hunk 10]
```diff
@@ -9,12 +9,14 @@
 	var/input_power_multiplier = 1
 
 	component_types = list(
-		///obj/item/weapon/circuitboard/tesla_coil,
-		/obj/item/weapon/stock_parts/capacitor
+		/obj/item/weapon/circuitboard/tesla_coil,
+		/obj/item/weapon/stock_parts/capacitor = 1
```
Comment (Lohikar) [inline code/modules/power/tesla/coil.dm]: `= 1` is pointless, just omit the number if it's 1.

File: code/modules/research/stockparts.dm [hunk 11]
```diff
@@ -0,0 +1,189 @@
+/obj/item/weapon/stock_parts
+	name = "stock part"
+	desc = "What?"
+	gender = PLURAL
+	icon = 'icons/obj/stock_parts.dmi'
+	w_class = 2.0
+	var/rating = 1
+
+/obj/item/weapon/stock_parts/New()
```
Comment (Lohikar) [inline code/modules/research/stockparts.dm]: Initialize.

**Comment by Lohikar** [comment]
I also don't really think upgradable APCs is worth it, especially if it requires those components to build them in the first place.

**Comment by printer16** [comment]
Reopened and untagged feedback required considering the change that required feedback is no longer in the PR.

File: code/game/machinery/kitchen/cooking_machines/_appliance.dm [hunk 12]
```diff
@@ -225,10 +235,17 @@
 		return
 
 	var/result = can_insert(I, user)
-	if (!result)
-		return
+	if(!result)
+		if(isscrewdriver(I))
```
Comment (Lohikar) [inline code/game/machinery/kitchen/cooking_machines/_appliance.dm]: Isn't there a helper for this on the base machinery type?

File: code/game/machinery/vending.dm [hunk 13]
```diff
@@ -53,6 +53,8 @@
 	var/active = 1 //No sales pitches if off!
 	var/vend_ready = 1 //Are we ready to vend?? Is it time??
 	var/vend_delay = 10 //How long does it take to vend?
+	var/inital_vend_delay = 10 //How long does it take to vend?
```
Comment (Lohikar) [inline code/game/machinery/vending.dm]: Might as well remove this var if it's not used.

File: code/modules/mining/machine_processing.dm [hunk 14]
```diff
@@ -298,3 +306,32 @@
 			continue
 
 	console.updateUsrDialog()
+
+/obj/machinery/mineral/processing_unit/attackby(obj/item/weapon/W as obj, mob/user as mob)
+	if(isscrewdriver(W))
```
Comment (Lohikar) [inline code/modules/mining/machine_processing.dm]: Isn't there a helper for this?

**Comment by printer16** [comment]
Updated.

**Pull request merged**